Sources/ComposableArchitecture/Effects/Cancellation.swift

```diff
@@ -26,53 +108,63 @@ extension Effect {
   ///   - cancelInFlight: Determines if any in-flight effect with the same identifier should be
   ///     canceled before starting this new one.
   /// - Returns: A new effect that is capable of being canceled by an identifier.
-  public func cancellable(id: AnyHashable, cancelInFlight: Bool = false) -> Effect {
+  public func cancellable(id: AnyHashable, cancelInFlight: Bool = false, bag: CancellationBag = .global) -> Effect {
     let effect = Deferred { () -> Publishers.HandleEvents<PassthroughSubject<Output, Failure>> in
-      cancellablesLock.lock()
-      defer { cancellablesLock.unlock() }
-
+      cancellablesLock.handle.lock()
+      defer { cancellablesLock.handle.unlock() }
+      
       let subject = PassthroughSubject<Output, Failure>()
       let cancellable = self.subscribe(subject)
-
+      
       var cancellationCancellable: AnyCancellable!
       cancellationCancellable = AnyCancellable {
-        cancellablesLock.sync {
+        cancellablesLock.sync({
           subject.send(completion: .finished)
           cancellable.cancel()
-          cancellationCancellables[id]?.remove(cancellationCancellable)
-          if cancellationCancellables[id]?.isEmpty == .some(true) {
-            cancellationCancellables[id] = nil
+          bag.cancellationCancellables[id]?.remove(cancellationCancellable)
+          if bag.cancellationCancellables[id]?.isEmpty == .some(true) {
+            bag.cancellationCancellables[id] = nil
           }
-        }
+        })
       }
-
-      cancellationCancellables[id, default: []].insert(
+      
+      bag.cancellationCancellables[id, default: []].insert(
         cancellationCancellable
       )
-
+      
       return subject.handleEvents(
         receiveCompletion: { _ in cancellationCancellable.cancel() },
         receiveCancel: cancellationCancellable.cancel
       )
     }
     .eraseToEffect()
-
+    
     return cancelInFlight ? .concatenate(.cancel(id: id), effect) : effect
   }
-
+  
   /// An effect that will cancel any currently in-flight effect with the given identifier.
   ///
   /// - Parameter id: An effect identifier.
   /// - Returns: A new effect that will cancel any currently in-flight effect with the given
   ///   identifier.
-  public static func cancel(id: AnyHashable) -> Effect {
+  public static func cancel(id: AnyHashable, bag: CancellationBag = .global) -> Effect {
     return .fireAndForget {
-      cancellablesLock.sync {
-        cancellationCancellables[id]?.forEach { $0.cancel() }
-      }
+      bag.cancel(id: id)
+    }
+  }
+  
+  public static func cancelAll(bag: CancellationBag) -> Effect {
+    return .fireAndForget {
+      bag.cancelAll()
     }
   }
 }
 
-var cancellationCancellables: [AnyHashable: Set<AnyCancellable>] = [:]
-let cancellablesLock = NSRecursiveLock()
+fileprivate struct Lock {
```

Comment:
TCA already provides an extension on NSRecursiveLock for sync located in Internal/Locking.swift

Reply:
Yup, just wanted a simple standalone file that could be copied out and in to any project